### ISSUE TITLE
gil_load.so: undefined symbol: clock_gettime

### ISSUE THREAD

**Issue by lsmithso** [chrisjbillington/gil_load#1]
I get this error when importing gil_load on python 2.7, either after a pip install or build from src. It works fine on python 3.

Python 2.7.3 (default, Jun 22 2015, 19:33:41) 
[GCC 4.6.3] on linux2
Type "help", "copyright", "credits" or "license" for more information.
> import gil_load 
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/lsmithso/junk/gil_load/env/local/lib/python2.7/site-packages/gil_load/__init__.py", line 1, in <module>
    from .gil_load import init, test, start, stop, get
ImportError: /home/lsmithso/junk/gil_load/env/local/lib/python2.7/site-packages/gil_load/gil_load.so: undefined symbol: clock_gettime

**Comment by chrisjbillington** [comment]
Thanks for the report. `That's` interesting - I would have expected failure at compile time rather than run time for that.

It looks like `clock_gettime` was added to the `posix.time` includes in cython version 0.21, released in July 2014. Is the version of cython you have in python 2 older than that? I notice your Python version (2.7.3) is pretty old, from 2012.

I could modify `gil_load.pyx` to `cdef extern` all the stuff it needs from `posix.time` so that it can use it even with older versions of Cython, but would rather not. What if I just `cdef extern` that one function? Does the following change to `gil_load.pyx` fix the problem for you?

```diff
diff --git a/gil_load/gil_load.pyx b/gil_load/gil_load.pyx
index e88bb92..2e933b4 100644
--- a/gil_load/gil_load.pyx
+++ b/gil_load/gil_load.pyx
@@ -11,7 +11,12 @@ from libc.stdio cimport printf, fprintf, FILE, fdopen, fflush
 from libc.string cimport memcpy
 from libc.math cimport log
 from libc.time cimport time, time_t, localtime, strftime, tm
-from posix.time cimport timespec, clockid_t, clock_gettime, CLOCK_MONOTONIC
+from posix.time cimport timespec, clockid_t, CLOCK_MONOTONIC
+
+
+cdef extern from "time.h" nogil:
+    timespec clock_gettime(clockid_t, timespec *)
+
 
 cdef extern from "pthread.h" nogil:
```

I've also considered uploading prebuild binaries, but this is tricky since I'd have to upload one for each combination of Python 2 vs 3 and 32 bit vs 64 bit, and I'm not sure what else - and it seems binary distributions for linux are not common practice.

**Comment by chrisjbillington** [comment]
Actually it looks like cython >= 0.22 is required.

You can upgrade cython with `pip install --upgrade cython`.

I've made the installer check for a new enough version of cython - automatically upgrading cython with requirements.txt or 'install_requires' doesn't seem to quite work if there is a system-installed cython already, due to some path issue I don't understand, so I've just made the installer prompt the user to do so.

**Comment by chrisjbillington** [comment]
Actually I've changed it now so that cython is not required for installation - so that the version installed from PyPI includes cythonised .c file for the extension already.

If you `pip install --upgrade gil_load`, hopefully it should be fixed regardless of what version of cython you have

**Comment by lsmithso** [comment]
Hi: Thanks for your swift response.

cython.__version__
'0.26'

I installed python 2.7.6

Python 2.7.6 (default, Aug 23 2017, 20:40:17) 
[GCC 4.6.3] on linux2
Type "help", "copyright", "credits" or "license" for more information.

And did both a pip install gil_load and a build from the repo clone
but got the same error as before:

ImportError: ./gil_load.so: undefined symbol: clock_gettime


I applied your patch but got exactly the same error again.



Chris Billington writes:
 > Thanks for the report. `That's` interesting - I would have expected failure at compile time rather than run time for that.
 > 
 > It looks like `clock_gettime` was added to the `posix.time` includes in cython version 0.21, released in July 2014. Is the version of cython you have in python 2 older than that? I notice your Python version (2.7.3) is pretty old, from 2012.
 > 
 > I could modify `gil_load.pyx` to `cdef extern` all the stuff it needs from `posix.time` so that it can use it even with older versions of Cython, but would rather not. What if I just `cdef extern` that one function? Does the following change to `gil_load.pyx` fix the problem for you?
 > 
 > ```diff
 > diff --git a/gil_load/gil_load.pyx b/gil_load/gil_load.pyx
 > index e88bb92..2e933b4 100644
 > --- a/gil_load/gil_load.pyx
 > +++ b/gil_load/gil_load.pyx
 > @@ -11,7 +11,12 @@ from libc.stdio cimport printf, fprintf, FILE, fdopen, fflush
 >  from libc.string cimport memcpy
 >  from libc.math cimport log
 >  from libc.time cimport time, time_t, localtime, strftime, tm
 > -from posix.time cimport timespec, clockid_t, clock_gettime, CLOCK_MONOTONIC
 > +from posix.time cimport timespec, clockid_t, CLOCK_MONOTONIC
 > +
 > +
 > +cdef extern from "time.h" nogil:
 > +    timespec clock_gettime(clockid_t, timespec *)
 > +
 >  
 >  cdef extern from "pthread.h" nogil:
 > ```
 > 
 > I've also considered uploading prebuild binaries, but this is tricky since I'd have to upload one for each combination of Python 2 vs 3 and 32 bit vs 64 bit, and I'm not sure what else - and it seems binary distributions for linux are not common practice.
 > 
 > -- 
 > You are receiving this because you authored the thread.
 > Reply to this email directly or view it on GitHub:
 > https://github.com/chrisjbillington/gil_load/issues/1#issuecomment-324402003

-- 
Les Smithson

**Comment by lsmithso** [comment]
Hi: I think our emails crossed.

I tried this anyway and got the same result.

Googling around I see that -lrt needs to be added to the linker. I'm
not familiar with cython, is it possible to add that lib to setup.py?



Chris Billington writes:
 > Actually I've changed it now so that cython is not required for installation - so that the version installed from PyPI includes cythonised .c file for the extension already.
 > 
 > If you `pip install --upgrade gil_load`, hopefully it should be fixed regardless of what version of cython you have
 > 
 > -- 
 > You are receiving this because you authored the thread.
 > Reply to this email directly or view it on GitHub:
 > https://github.com/chrisjbillington/gil_load/issues/1#issuecomment-324455388

-- 
Les Smithson

**Comment by chrisjbillington** [comment]
Ah, I see!

Yes it is possible - in setup.py you can add 'extra_compile_args' to the `Extension` instance something like:

```python
extra_compile_args = ["-lrt"]

ext_modules = [Extension('gil_load.gil_load',
                         ["gil_load/gil_load" + ext],
                         extra_compile_args=extra_compile_args)]
```

I can easily add this to setup.py if it's the solution. Could you point me to some of the google results that made you suspect this was the solution? I'm just curious why that's required on your system but not on the ones I've used.

**Comment by chrisjbillington** [comment]
Ah, don't worry, I found it. According to this [stackoverflow thread](https://stackoverflow.com/questions/2418157/ubuntu-linux-c-error-undefined-reference-to-clock-gettime-and-clock-settim), `-lrt` is required for glibc < 2.17, whereas later versions include it in the standard library.

I've included the fix and released on PyPI as version 0.3.4

**Comment by lsmithso** [comment]
Hi:

https://stackoverflow.com/questions/2418157/ubuntu-linux-c-error-undefined-reference-to-clock-gettime-and-clock-settim 


Chris Billington writes:
 > Ah, I see!
 > 
 > Yes it is possible - in setup.py you can add 'extra_compile_args' to the `Extension` instance something like:
 > 
 > ```python
 > extra_compile_args = ["-lrt"]
 > 
 > ext_modules = [Extension('gil_load.gil_load',
 >                          ["gil_load/gil_load" + ext],
 >                          extra_compile_args=extra_compile_args)]
 > ```
 > 
 > I can easily add this to setup.py if it's the solution. Could you point me to some of the google results that made you suspect this was the solution? I'm just curious why that's required on your system but not on the ones I've used.
 > 
 > -- 
 > You are receiving this because you authored the thread.
 > Reply to this email directly or view it on GitHub:
 > https://github.com/chrisjbillington/gil_load/issues/1#issuecomment-324463260

-- 
Les Smithson

**Comment by chrisjbillington** [comment]
And the reason this was not a problem for you in python 3 is that CPython itself links against librt since CPython 3.5

https://github.com/python/cpython/commit/7efb83393cbe08924682c6852e94d3c4a4033c34

**Comment by lsmithso** [comment]
Hi: I tried 0.3.4 and got the same error. I'm not sure its building
correctly:

This doesn't include librt.so in the link:

gcc -pthread -shared build/temp.linux-x86_64-2.7/gil_load/gil_load.o -o build/lib.linux-x86_64-2.7/gil_load/gil_load.so

and:

ldd build/lib.linux-x86_64-2.7/gil_load/gil_load.so
	linux-vdso.so.1 =>  (0x00007fff1db5a000)
	libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f1d299e6000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f1d29628000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f1d29e3a000)

	Doesn't show it either.

	Linking by hand:

	gcc -pthread -shared
	build/temp.linux-x86_64-2.7/gil_load/gil_load.o -lrt -o
	build/lib.linux-x86_64-2.7/gil_load/gil_load.so

	Links in librt correctly and allows me to import gil_load
	successfully . We are getting close!

ldd build/lib.linux-x86_64-2.7/gil_load/gil_load.so
	linux-vdso.so.1 =>  (0x00007fff9527f000)
	librt.so.1 => /lib/x86_64-linux-gnu/librt.so.1 (0x00007f155199a000)
	libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f155177d000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f15513be000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f1551dd9000)

	Is it a problem with my version of setuptools?

	setuptools.__version__
'36.2.7'
> 
	
Chris Billington writes:
 > And the reason this was not a problem for you in python 3 is that CPython itself links against librt since CPython 3.5
 > 
 > https://github.com/python/cpython/commit/7efb83393cbe08924682c6852e94d3c4a4033c34
 > 
 > -- 
 > You are receiving this because you authored the thread.
 > Reply to this email directly or view it on GitHub:
 > https://github.com/chrisjbillington/gil_load/issues/1#issuecomment-324468970

-- 
Les Smithson

**Comment by lsmithso** [comment]
Hi: With the aforementioned hack I'm getting some really useful
results.  Just to be clear, the result is the fraction of time that
the GIL is held. Is it possible to measure the fraction of time that
threads spend waiting for the GIL?


Chris Billington writes:
 > And the reason this was not a problem for you in python 3 is that CPython itself links against librt since CPython 3.5
 > 
 > https://github.com/python/cpython/commit/7efb83393cbe08924682c6852e94d3c4a4033c34
 > 
 > -- 
 > You are receiving this because you authored the thread.
 > Reply to this email directly or view it on GitHub:
 > https://github.com/chrisjbillington/gil_load/issues/1#issuecomment-324468970

-- 
Les Smithson

**Comment by chrisjbillington** [comment]
Good to hear it's useful!

Yes, the first number printed to file or returned by `get()` is the fraction of the time the GIL was held, and the other three numbers are 1, 5, and 15 minute exponential moving averages of the same.

A caveat regarding accuracy though: the fraction of time the GIL is held should be treated as having a fractional uncertainty of about sqrt(av_sample_interval / duration_of_profiling), since the measuring is done by sampling the GIL at random intervals to see if it is held, which will be statistically inaccurate if only few samples are taken.

Unfortunately measuring threads being blocked waiting for the GIL would be much more involved! Perhaps one day when I'm feeling like going down that rabbit hole, but for now just polling the GIL to see how often it is held is much simpler.

Ah, I didn't realise setuptools would compile and link with separate calls to gcc. If you pass in `extra_link_args` instead of `extra_compile_args` in `setup.py`, does that make setuptools call the right linker command?

```diff

diff --git a/setup.py b/setup.py
index b17b1d9..56a9f5b 100644
--- a/setup.py
+++ b/setup.py
@@ -29,11 +29,11 @@ else:
     ext = '.c'
 
 # Required for glibc < 2.17:
-extra_compile_args = ["-lrt"]
+extra_link_args = ["-lrt"]
 
 ext_modules = [Extension('gil_load.gil_load',
                          ["gil_load/gil_load" + ext],
-                         extra_compile_args=extra_compile_args)]
+                         extra_link_args=extra_link_args)]
 
 if USE_CYTHON:
     from Cython.Build import cythonize
```

For me this produces the following gcc command for the linking step when I run `python setup.py build_ext`:

`x86_64-linux-gnu-gcc -pthread -shared -Wl,-O1 -Wl,-Bsymbolic-functions -Wl,-Bsymbolic-functions -Wl,-z,relro -fno-strict-aliasing -DNDEBUG -g -fwrapv -O2 -Wall -Wstrict-prototypes -D_FORTIFY_SOURCE=2 -g -fstack-protector --param=ssp-buffer-size=4 -Wformat -Werror=format-security build/temp.linux-x86_64-2.7/gil_load/gil_load.o -o build/lib.linux-x86_64-2.7/gil_load/gil_load.so -lrt`

So I see the "-lrt" at the end, which was previously in the compile step where I suppose it had no effect.

**Comment by lsmithso** [comment]
Hi: That worked!  Thanks very much for this.

Chris Billington writes:
 > Good to hear it's useful!
 > 
 > Yes, the first number printed to file or returned by `get()` is the fraction of the time the GIL was held, and the other three numbers are 1, 5, and 15 minute exponential moving averages of the same.
 > 
 > A caveat regarding accuracy though: the fraction of time the GIL is held should be treated as having a fractional uncertainty of about sqrt(av_sample_interval / duration_of_profiling), since the measuring is done by sampling the GIL at random intervals to see if it is held, which will be statistically inaccurate if only few samples are taken.
 > 
 > Unfortunately measuring threads being blocked waiting for the GIL would be much more involved! Perhaps one day when I'm feeling like going down that rabbit hole, but for now just polling the GIL to see how often it is held is much simpler.
 > 
 > Ah, I didn't realise setuptools would compile and link with separate calls to gcc. If you pass in `extra_link_args` instead of `extra_compile_args` in `setup.py`, does that make setuptools call the right linker command?
 > 
 > ```diff
 > 
 > diff --git a/setup.py b/setup.py
 > index b17b1d9..56a9f5b 100644
 > --- a/setup.py
 > +++ b/setup.py
 > @@ -29,11 +29,11 @@ else:
 >      ext = '.c'
 >  
 >  # Required for glibc < 2.17:
 > -extra_compile_args = ["-lrt"]
 > +extra_link_args = ["-lrt"]
 >  
 >  ext_modules = [Extension('gil_load.gil_load',
 >                           ["gil_load/gil_load" + ext],
 > -                         extra_compile_args=extra_compile_args)]
 > +                         extra_link_args=extra_link_args)]
 >  
 >  if USE_CYTHON:
 >      from Cython.Build import cythonize
 > ```
 > 
 > For me this produces the following gcc command for the linking step when I run `python setup.py build_ext`:
 > 
 > `x86_64-linux-gnu-gcc -pthread -shared -Wl,-O1 -Wl,-Bsymbolic-functions -Wl,-Bsymbolic-functions -Wl,-z,relro -fno-strict-aliasing -DNDEBUG -g -fwrapv -O2 -Wall -Wstrict-prototypes -D_FORTIFY_SOURCE=2 -g -fstack-protector --param=ssp-buffer-size=4 -Wformat -Werror=format-security build/temp.linux-x86_64-2.7/gil_load/gil_load.o -o build/lib.linux-x86_64-2.7/gil_load/gil_load.so -lrt`
 > 
 > So I see the "-lrt" at the end, which was previously in the compile step where I suppose it had no effect.
 > 
 > -- 
 > You are receiving this because you authored the thread.
 > Reply to this email directly or view it on GitHub:
 > https://github.com/chrisjbillington/gil_load/issues/1#issuecomment-324708442

-- 
Les Smithson

**Comment by lsmithso** [comment]
Hi: That worked!  Thanks very much for the fix.

Chris Billington writes:
 > Good to hear it's useful!
 > 
 > Yes, the first number printed to file or returned by `get()` is the fraction of the time the GIL was held, and the other three numbers are 1, 5, and 15 minute exponential moving averages of the same.
 > 
 > A caveat regarding accuracy though: the fraction of time the GIL is held should be treated as having a fractional uncertainty of about sqrt(av_sample_interval / duration_of_profiling), since the measuring is done by sampling the GIL at random intervals to see if it is held, which will be statistically inaccurate if only few samples are taken.
 > 
 > Unfortunately measuring threads being blocked waiting for the GIL would be much more involved! Perhaps one day when I'm feeling like going down that rabbit hole, but for now just polling the GIL to see how often it is held is much simpler.
 > 
 > Ah, I didn't realise setuptools would compile and link with separate calls to gcc. If you pass in `extra_link_args` instead of `extra_compile_args` in `setup.py`, does that make setuptools call the right linker command?
 > 
 > ```diff
 > 
 > diff --git a/setup.py b/setup.py
 > index b17b1d9..56a9f5b 100644
 > --- a/setup.py
 > +++ b/setup.py
 > @@ -29,11 +29,11 @@ else:
 >      ext = '.c'
 >  
 >  # Required for glibc < 2.17:
 > -extra_compile_args = ["-lrt"]
 > +extra_link_args = ["-lrt"]
 >  
 >  ext_modules = [Extension('gil_load.gil_load',
 >                           ["gil_load/gil_load" + ext],
 > -                         extra_compile_args=extra_compile_args)]
 > +                         extra_link_args=extra_link_args)]
 >  
 >  if USE_CYTHON:
 >      from Cython.Build import cythonize
 > ```
 > 
 > For me this produces the following gcc command for the linking step when I run `python setup.py build_ext`:
 > 
 > `x86_64-linux-gnu-gcc -pthread -shared -Wl,-O1 -Wl,-Bsymbolic-functions -Wl,-Bsymbolic-functions -Wl,-z,relro -fno-strict-aliasing -DNDEBUG -g -fwrapv -O2 -Wall -Wstrict-prototypes -D_FORTIFY_SOURCE=2 -g -fstack-protector --param=ssp-buffer-size=4 -Wformat -Werror=format-security build/temp.linux-x86_64-2.7/gil_load/gil_load.o -o build/lib.linux-x86_64-2.7/gil_load/gil_load.so -lrt`
 > 
 > So I see the "-lrt" at the end, which was previously in the compile step where I suppose it had no effect.
 > 
 > -- 
 > You are receiving this because you authored the thread.
 > Reply to this email directly or view it on GitHub:
 > https://github.com/chrisjbillington/gil_load/issues/1#issuecomment-324708442

-- 
Les Smithson

**Comment by lsmithso** [comment]
Hi: It worked! Thanks very much for your help.

Chris Billington writes:
 > Good to hear it's useful!
 > 
 > Yes, the first number printed to file or returned by `get()` is the fraction of the time the GIL was held, and the other three numbers are 1, 5, and 15 minute exponential moving averages of the same.
 > 
 > A caveat regarding accuracy though: the fraction of time the GIL is held should be treated as having a fractional uncertainty of about sqrt(av_sample_interval / duration_of_profiling), since the measuring is done by sampling the GIL at random intervals to see if it is held, which will be statistically inaccurate if only few samples are taken.
 > 
 > Unfortunately measuring threads being blocked waiting for the GIL would be much more involved! Perhaps one day when I'm feeling like going down that rabbit hole, but for now just polling the GIL to see how often it is held is much simpler.
 > 
 > Ah, I didn't realise setuptools would compile and link with separate calls to gcc. If you pass in `extra_link_args` instead of `extra_compile_args` in `setup.py`, does that make setuptools call the right linker command?
 > 
 > ```diff
 > 
 > diff --git a/setup.py b/setup.py
 > index b17b1d9..56a9f5b 100644
 > --- a/setup.py
 > +++ b/setup.py
 > @@ -29,11 +29,11 @@ else:
 >      ext = '.c'
 >  
 >  # Required for glibc < 2.17:
 > -extra_compile_args = ["-lrt"]
 > +extra_link_args = ["-lrt"]
 >  
 >  ext_modules = [Extension('gil_load.gil_load',
 >                           ["gil_load/gil_load" + ext],
 > -                         extra_compile_args=extra_compile_args)]
 > +                         extra_link_args=extra_link_args)]
 >  
 >  if USE_CYTHON:
 >      from Cython.Build import cythonize
 > ```
 > 
 > For me this produces the following gcc command for the linking step when I run `python setup.py build_ext`:
 > 
 > `x86_64-linux-gnu-gcc -pthread -shared -Wl,-O1 -Wl,-Bsymbolic-functions -Wl,-Bsymbolic-functions -Wl,-z,relro -fno-strict-aliasing -DNDEBUG -g -fwrapv -O2 -Wall -Wstrict-prototypes -D_FORTIFY_SOURCE=2 -g -fstack-protector --param=ssp-buffer-size=4 -Wformat -Werror=format-security build/temp.linux-x86_64-2.7/gil_load/gil_load.o -o build/lib.linux-x86_64-2.7/gil_load/gil_load.so -lrt`
 > 
 > So I see the "-lrt" at the end, which was previously in the compile step where I suppose it had no effect.
 > 
 > -- 
 > You are receiving this because you authored the thread.
 > Reply to this email directly or view it on GitHub:
 > https://github.com/chrisjbillington/gil_load/issues/1#issuecomment-324708442

-- 
Les Smithson

**Comment by chrisjbillington** [comment]
Fixed in 36de61d8f released as 0.3.5 on PyPI.

Also I've figured our how one might track when threads are blocking waiting for the GIL, with some LD_PRELOAD trickery to intercept calls to `sem_wait` in Python 2 and `pthread_mutex_lock`/`pthread_mutex_unlock` in Python 3 - these are used in CPython to hold a different lock (not the GIL itself) whilst a thread is waiting for the GIL. I'm busy writing a thesis at the moment but I will probably implement this in the future!
